First version of the Couchbase Cloud client SDK

zain-b opened the pull request:
# couchbase-cloud-go-client
Work in progress Go wrapper for the Couchbase Cloud REST API

## Usage

`go get github.com/couchbaselabs/couchbase-cloud-go-client`

### Import
```go
import (
    "github.com/couchbaselabs/couchbase-cloud-go-client"
)
```

### Client

```go
client := couchbasecloudclient.NewClient(os.Getenv("CBC_ACCESS_KEY"), os.Getenv("CBC_SECRET_KEY"))
```

### List all clouds
Options can be specified with the `ListCloudOptions` type (some are optional)
```go
page := 1
lastPage := math.MaxInt16

for ok := true; ok; ok = page <= lastPage {
    listCloudsResponse, err := client.ListClouds(&couchbasecloudclient.ListCloudsOptions{Page: page, PerPage: 10})

    if err != nil {
        return nil, err
    }

    for _, cloud := range listCloudsResponse.Data {
        fmt.Printf(cloud.Name)
    }

    lastPage = listCloudsResponse.Cursor.Pages.Last
    page++
}
```

### List all clusters
Options can be specified with the `ListClustersOptions` type (some are optional)
```go
page := 1
lastPage := math.MaxInt16

for ok := true; ok; ok = page <= lastPage {
    listClustersResponse, err := client.ListClusters(&couchbasecloudclient.ListClustersOptions{Page: page, PerPage: 10})

    if err != nil {
        return nil, err
    }

    for _, cluster := range listClustersResponse.Data {
        fmt.Printf(cluster.Name)
    }

    lastPage = listClustersResponse.Cursor.Pages.Last
    page++
}
```

## TODO

- Build better (functional) pagination capabilities into the client
- Unit/integration testing

cg14823 commented on client.go:
```diff
@@ -0,0 +1,111 @@
+const couchbaseCloudBaseUrl = "https://cloudapi.cloud.couchbase.com"
+const requestTimeoutMinutes = 5
+
+type CouchbaseCloudClient struct {
```
If this is going to be a publicly or even internal available SDK I suggest adding doc strings to all Public types functions so that documentation can be auto-generated from the code.

cg14823 commented on client.go:
```diff
@@ -0,0 +1,111 @@
+		return fmt.Errorf("unknown error, status code: %d", res.StatusCode)
+	}
+
+	if err = json.NewDecoder(res.Body).Decode(&dataType); err != nil {
```
for brevity this can be shortened to just:

```
return json.NewDecoder(res.Body).Decode(&dataType)
```

cg14823 commented on clusters.go:
```diff
@@ -0,0 +1,80 @@
+	Data   []Cluster `json:"data"`
+}
+
+type Cluster struct {
```
you have annotations for the ClusterList one but not Cluster not sure if that is intentional or not but it may be useful to have annotations as the default marshalling in go would use capital letters and CamelCase which is not very pretty in JSON.

cg14823 commented on clouds.go:
```diff
@@ -0,0 +1,70 @@
+
+const listCloudsUrl = "/clouds"
+
+func (client *CouchbaseCloudClient) ListClouds(options *ListCloudsOptions) (*CloudsList, error) {
```
It be nice to have versions with iterator functions kind of like aws has https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.ListObjectsV2Pages, I know you mention it is something you where planing to do but though might as well mention it just in case :smile:

cg14823 reviewed (commented):
Hi, awesome work, and I think this could be very useful as it grows. I would recommend adding unit tests (although it may require mocking a bit of the API) as well as getting some CV setup and maybe running some of https://github.com/golangci/golangci-lint which helps maintain code quality.

cg14823 commented:
If the cloud API as an open spec file that is up to date maybe it can be used to auto-generate a large amount of the client as well.

zain-b commented:
> Hi, awesome work, and I think this could be very useful as it grows. I would recommend adding unit tests (although it may require mocking a bit of the API) as well as getting some CV setup and maybe running some of https://github.com/golangci/golangci-lint which helps maintain code quality.

Thanks a lot for the comments @cg14823, I've addressed some of them and added some of the longer term suggestions into the TODO section of the README 😃 

This will be merged into main and I'll create an alpha release to support the tools currently depending on it
